RFC: Reorganize settings.

artcz opened the pull request:
Main goal here is to allow everything except secrets to be under version control (so no more unversioned settings_locale.py)

Mostly by splitting settings into different modules based on where they're used (like dev.py, production.py and automated_testing.py) also introducing python-decouple, pathlib2, and dj-database-url

The goal here is to have exactly the same behaviour, but easier to update/modify. There is more detailed description in each commit, also reading commit by commit will make the diff easier to understand.

artcz committed:
Split INSTALLED_APPS to DJANGO/THIRD_PARTY/PROJECT_APPS for readability

artcz committed:
settings.py: fix indent in DATABASE settings; remove unused LESS_THAN_1{7,8}

artcz committed:
Clean up settings by introducing python-decouple, pathlib and dj-database-url

To simplify settings a bit, I started a series of commits that aim to make the
settings easier to understand and manage. Eventually I want to make it such
that we will have everything (except secrets) under version control (so no more
unversioned settings_locale.py)

This first commit introduces python-decouple, pathlib2 (because python2.7) and
dj_database_url to clean up the secrets/environment variables situation

Instead of using os.environ to get the secrets it's using more reliable
decouple package, that has builtin casting support (so for example 'False' is
False not True).

artcz committed:
Move pycon/settings.py to pycon/settings/core.py.

In this commit I replaced the setup that used unversioned settings_locale.py
with a separate files for different environments, all grouped together in new
directory called pycon/settings.

Changing which settings file to use could be achieved by exporting
DJANGO_SETTINGS_MODULE (for example giving it value 'pycon.settings.dev') or
modifying the default values in manage.py (settings.dev), wsgi.py
(settings.production) or pytest.ini (settings.automated_testing)

From this point we can store all the settings in versioned files, just keeping
secrets in the unversioned .env file.

artcz committed:
Fixed cron_lookup path related to new settings.py setup and location

artcz committed:
Fixed the missing django_admin_style skin in django admin

artcz committed:
Added explicit python2.7 to provision.sh

artcz committed:
Update order of pages in initial data

Looks like the first page is homepage (not the last one), reflecting that
change in this update. Also making them published so they are visible for
unauthenticated users as well.

artcz committed:
Fixed .env handling on travis

malemburg reviewed (changes requested):
* pycon/settings/local.py should not be versioned
* please keep the local.py file instead of moving env settings to .env; storing env settings in a hidden file is not a good idea
* Travis env should not include our keys (in fact, these should be moved to the OS env and passed in via docker)

artcz committed:
Added additional local_dev.py that inherits from dev.py

Also added manage_production.py that automatically sets DJANGO_SETTINGS_MODULE
to pycon.settings.production

malemburg commented:
I don't like this PR. It's a classical case of over-engineering a very simple thing: configuring a Django app.

There are also too catchy details in there such as manage.py defaulting to dev settings, while wsgi,py default to production ones.

We don't need all this.

If you're developing, you can easily have a custom setting_local.py in your work dir to run in dev mode or whatever other mode you prefer.

I understand that for CI testing, some settings need to fixed (e.g. sending emails should not happen), but this can easily be had with a custom settings module for Travis. No need to complicate the rest of the code because of this.

Some other things I noted before I stopped the review:

- Please don't use relative imports.
- Defaulting wsgi.py to production settings is not a good idea.
- Please remove .env and put the .env.travis definitions into travis.yml
- Documentation about available env vars can go into the README
- The local.py settings should be sourced as settings regardless of which mode 

Let's please focus on real issues we have with the code.
